sphinx/directives/other.py

```diff
@@ -361,6 +364,92 @@ def run(self) -> List[Node]:
+        path = os.path.normpath(os.path.join(source_dir, path))
+        path = relative_path(None, path)
+        path = nodes.reprunicode(path)
+        encoding = self.options.get(
+            'encoding', self.state.document.settings.input_encoding)
+        e_handler = self.state.document.settings.input_encoding_error_handler
+        try:
+            self.state.document.settings.record_dependencies.add(path)
+            include_file = io.FileInput(source_path=path,
+                                        encoding=encoding,
+                                        error_handler=e_handler)
+        except UnicodeEncodeError:
+            raise self.severe(u'Problems with "%s" directive path:\n'
+                              'Cannot encode input file path "%s" '
+                              '(wrong locale?).' %
+                              (self.name, SafeString(path)))
+        except IOError as error:
+            raise self.severe(u'Problems with "%s" directive path:\n%s.' %
+                              (self.name, ErrorString(error)))
+        startline = self.options.get('start-line', None)
+        endline = self.options.get('end-line', None)
+        try:
+            if startline or (endline is not None):
+                lines = include_file.readlines()
+                rawtext = ''.join(lines[startline:endline])
+            else:
+                rawtext = include_file.read()
+        except UnicodeError as error:
+            raise self.severe(u'Problem with "%s" directive:\n%s' %
+                              (self.name, ErrorString(error)))
+        # start-after/end-before: no restrictions on newlines in match-text,
+        # and no restrictions on matching inside lines vs. line boundaries
+        after_text = self.options.get('start-after', None)
+        if after_text:
+            # skip content in rawtext before *and incl.* a matching text
+            after_index = rawtext.find(after_text)
+            if after_index < 0:
+                raise self.severe('Problem with "start-after" option of "%s" '
+                                  'directive:\nText not found.' % self.name)
+            rawtext = rawtext[after_index + len(after_text):]
+        before_text = self.options.get('end-before', None)
+        if before_text:
+            # skip content in rawtext after *and incl.* a matching text
+            before_index = rawtext.find(before_text)
+            if before_index < 0:
+                raise self.severe('Problem with "end-before" option of "%s" '
+                                  'directive:\nText not found.' % self.name)
+            rawtext = rawtext[:before_index]
+
+        # copied code ends
+        app = self.env.app
+        filetype = get_filetype(app.config.source_suffix, path)
+        parser = app.registry.create_source_parser(app, filetype)
+
+        sub_document = new_document(path, self.state.document.settings)
+        parser.parse(rawtext, sub_document)
```

Comment:
Why do you generate new document here? It prevents to share local reference context (ex. footnotes, hyperref, etc.).

Reply:
Is there another way to invoke the parser?

I figured it didn't matter, because if the included document is in a different file it's very unlikely to reuse footnotes.

Reply:
>I figured it didn't matter, because if the included document is in a different file it's very unlikely to reuse footnotes.

The document object is large database for the doctree. So all references would be broken. Is it okay to conflict node_IDs between parent and included document? I guess `:ref:` role for included document may not work also.

>Is there another way to invoke the parser?

How about `parser.parse(rawtext, self.state.document)` instead?

Reply:
Does docutils allow a document to be parsed into multiple times? How does it know where the current insertion point is?